test/router.test.ts

```diff
@@ -0,0 +1,438 @@
+const FEE_COLLECTOR = RICH_WALLET_ADDR_0;
+
+const SIGNER_REFERRAL = 1;
+const WRAP_MODE_NONE = 0;
+const ZERO_ADDRESS = constants.ZERO_ADDRESS;
+const PAUSED = ZERO_ADDRESS;
+
+async function deployRouter(deployer: Deployer): Promise<Contract> {
+  const artifact = await deployer.loadArtifact('Router');
+  return await deployer.deploy(artifact, [
+    WRAPPED_NATIVE,
+    OWNER,
+    PAUSER,
+    FEE_COLLECTOR,
+    utils.hashBytecode(agentBytecode),
+  ]);
+}
+
+async function deployMockERC20(deployer: Deployer): Promise<Contract> {
+  const artifact = await deployer.loadArtifact('MockERC20');
+  return await deployer.deploy(artifact, ['mockERC20', 'mock']);
+}
+
+async function deployMockFallback(deployer: Deployer): Promise<Contract> {
+  const artifact = await deployer.loadArtifact('MockFallback');
+  return await deployer.deploy(artifact, []);
+}
+
+async function deployMockProtocol(deployer: Deployer, routerAddress): Promise<Contract> {
+  const artifact = await deployer.loadArtifact('MockProtocol');
+  return await deployer.deploy(artifact, [routerAddress]);
+}
+
+describe('Router', function () {
+  const inputsEmpty = [];
+  const logicsEmpty = [];
+  const logicBatchEmpty = { logics: [], fees: [], deadline: 0 };
+  const tokensReturnEmpty = [];
+
+  let wallet;
+  let provider;
+  let router;
+  let mockERC20;
+  let mockTo;
+  let mockProtocol;
+
+  async function setUpTest() {
+    provider = Provider.getDefaultProvider();
+    wallet = new Wallet(RICH_WALLET_PK_0, provider);
+    const deployer = new Deployer(hre, wallet);
+    router = await deployRouter(deployer);
+    mockERC20 = await deployMockERC20(deployer);
+    mockTo = await deployMockFallback(deployer);
+    mockProtocol = await deployMockProtocol(deployer, router.address);
+  }
+
+  before(async function () {});
+
+  beforeEach(async function () {});
+
+  afterEach(async function () {});
+
+  describe('Normal', function () {
+    beforeEach(async function () {
+      await setUpTest();
+    });
+
+    it('setup', async function () {
+      expect((await router.agentImplementation()) != ZERO_ADDRESS);
+      expect((await router.currentUser()) == INIT_CURRENT_USER);
+      expect((await router.pauser()) == PAUSER);
+      expect((await router.feeCollector()) == FEE_COLLECTOR);
+      expect((await router.owner()) == OWNER);
+    });
+
+    it('new agent', async function () {
+      // use sendTransaction here due to error msg: newAgent is not a function
+      const calldata = abi.simpleEncode('newAgent()');
+      const receipt = await wallet.sendTransaction({ to: router.address, data: calldata });
+      await receipt.wait();
+
+      const salt = rightPad(wallet.address.toString(), 64, '0'); // Turn wallet address to bytes32
+      const inputBytecode = abiCoder.encode(['address'], [await router.agentImplementation()]);
+      const newAgentAddr = utils.create2Address(router.address, utils.hashBytecode(agentBytecode), salt, inputBytecode);
```

Comment:
Why not getting the `newAgentAddr` from the return value?

Reply:
the tricky thing is if the return value is `address(0)`, this test case still passes if we use the return value.
I found using utils to verify the new agent address was more solid. What do you think?

Reply:
Not quite sure in what case the return value will be `address(0)` and not reverted. I think both of them make sense, but this test case is the only place that verifies if the return value is correct.

Reply:
woops! bad example in the above. I meant no matter what happens, the return value of `getAgent`(which reads the mapping `agents`) is always the same as the return value of `newAgent`. In my opinion, verifying the return value of `getAgent` is the same as verifying the return value of `newAgent`.

Reply:
Their value should be the same, but that's when the return value is correctly implemented. If the function is revised in the future, and the author forget to return the created agent address, this will be the test case to figure out that something is not working as expected.

Reply:
I see. Add a TODO here for now and will figure out how to compare the returned new agent address later.